nilearn/connectivity/connectivity_matrices.py

```diff
@@ -0,0 +1,341 @@
+
+    `robust_mean_` : numpy.ndarray
+        The mean connectivity for the robust dispersion kind.
+
+    `whitening_` : numpy.ndarray
+        The inverted square-rooted geometric mean of the covariance matrices.
+
+    References
+    ----------
+    For the use of "robust dispersion", see the paper:
+    G. Varoquaux et al. "Detection of brain functional-connectivity difference
+    in post-stroke patients using group-level covariance modeling, MICCAI 2010.
+    """
+
+    def __init__(self, cov_estimator=EmpiricalCovariance(assume_centered=True),
+                 kind='covariance'):
+        self.cov_estimator = cov_estimator
+        self.kind = kind
+
+    def fit(self, X, y=None):
+        """Fit the covariance estimator to the given time series for each
+        subject.
+
+        Parameters
+        ----------
+        X : list of numpy.ndarray, shapes (n_samples, n_features)
+            The input subjects time series.
+
+        Returns
+        -------
+        self : ConnectivityMatrix instance
+            The object itself. Useful for chaining operations.
+        """
+        self.cov_estimator_ = clone(self.cov_estimator)
+
+        if self.kind == 'robust dispersion':
+            covariances = [self.cov_estimator_.fit(x).covariance_ for x in X]
+            self.robust_mean_ = _geometric_mean(covariances, max_iter=30,
+                                                tol=1e-7)
+            self.whitening_ = _map_eigenvalues(lambda x: 1. / np.sqrt(x),
+                                               self.robust_mean_)
+
+        return self
+
+    def transform(self, X):
+        """Apply transform to covariances matrices to get the connectivity
+        matrices for the chosen kind.
+
+        Parameters
+        ----------
+        X : list of numpy.ndarray with shapes (n_samples, n_features)
+            The input subjects time series.
+
+        Returns
+        -------
+        output : numpy.ndarray, shape (n_samples, n_features, n_features)
+             The transformed connectivity matrices.
+        """
+        covariances = [self.cov_estimator_.fit(x).covariance_ for x in X]
+        covariances = np.array(covariances)
+        if self.kind == 'covariance':
+            connectivities = covariances
+        elif self.kind == 'robust dispersion':
+            connectivities = [_map_eigenvalues(np.log, self.whitening_.dot(
+                                               cov).dot(self.whitening_))
+                              for cov in covariances]
+        elif self.kind == 'precision':
+            connectivities = [linalg.inv(cov) for cov in covariances]
+        elif self.kind == 'partial correlation':
+            connectivities = [_prec_to_partial(linalg.inv(cov))
+                              for cov in covariances]
+        elif self.kind == 'correlation':
+            connectivities = [_cov_to_corr(cov) for cov in covariances]
+        else:
+            raise ValueError("Unknown connectivity kind.")
```

Comment:
It would be nice if the error message could mention both the allowed connectivity kinds and the one that the user provided.